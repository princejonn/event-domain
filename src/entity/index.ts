export * from "./Aggregate";
export * from "./Saga";
export * from "./View";
