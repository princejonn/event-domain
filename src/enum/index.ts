export * from "./EventField";
export * from "./EventStoreField";
export * from "./Evt";
export * from "./MessageBusType";
export * from "./MessageType";
export * from "./MetaState";
export * from "./SagaStoreField";
export * from "./StoreType";
export * from "./ViewStoreField";
