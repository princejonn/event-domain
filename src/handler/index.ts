export * from "./AggregateCommandHandler";
export * from "./AggregateEventHandler";
export * from "./SagaEventHandler";
export * from "./ViewEventHandler";
