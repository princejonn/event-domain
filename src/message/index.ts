export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
