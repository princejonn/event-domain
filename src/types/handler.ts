export interface HandlerIdentifier {
  name: string;
  context: string;
}

export interface HandlerConditions {
  created?: boolean;
  permanent?: boolean;
}

export interface HandlerIdentifierExpectingStructure {
  name?: string;
  context?: string;
}

export interface HandlerIdentifierMultipleContexts {
  name: string;
  context?: Array<string> | string;
}
