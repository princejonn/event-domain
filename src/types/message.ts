import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
