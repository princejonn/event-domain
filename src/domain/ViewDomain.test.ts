import MockDate from "mockdate";
import { DomainEvent } from "../message";
import { StoreType } from "../enum";
import { View } from "../entity";
import { ViewDomain } from "./ViewDomain";
import { ViewEventHandler } from "../handler";
import { logger } from "../test";
import {
  DomainError,
  HandlerNotRegisteredError,
  ViewAlreadyCreatedError,
  ViewNotCreatedError,
} from "../error";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("ViewDomain", () => {
  describe("when registering an event handler", () => {
    let domain: ViewDomain<any>;

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("eventName", ctx.event.name);
          },
        }),
      );
    });

    it("should become part of the domain", () => {
      // @ts-ignore
      expect(domain.eventHandlers).toMatchSnapshot();
    });

    it("should add subscriptions to topic", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].topic).toMatchSnapshot();
    });

    it("should add event subscription", async () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].sub).toMatchSnapshot();
    });

    it("should add handler callback", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].callback).toStrictEqual(
        expect.any(Function),
      );
    });
  });

  describe("when registering an existing event handler", () => {
    let domain: ViewDomain<any>;

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("eventName", ctx.event.name);
          },
        }),
      );
    });

    it("should throw", async () => {
      await expect(
        domain.registerEventHandler(
          new ViewEventHandler({
            aggregate: {
              name: "name",
              context: "context",
            },
            view: {
              name: "name",
              context: "context",
            },
            eventName: "eventName",
            getViewId: (event) => event.aggregate.id,
            handler: async (ctx) => {
              ctx.setState("eventName", ctx.event.name);
            },
          }),
        ),
      ).rejects.toThrow(Error);
    });

    it("should not register a second handler on domain", () => {
      // @ts-ignore
      expect(domain.eventHandlers).toMatchSnapshot();
    });
  });

  describe("when registering an event handler to multiple contexts", () => {
    let domain: ViewDomain<any>;

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: ["context", "secondary"],
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("eventName", ctx.event.name);
          },
        }),
      );
    });

    it("should become part of the domain for each context", () => {
      // @ts-ignore
      expect(domain.eventHandlers).toMatchSnapshot();
    });

    it("should add subscriptions to topic", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].topic).toMatchSnapshot();
    });

    it("should add event subscription", async () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].sub).toMatchSnapshot();
    });

    it("should add handler callback", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].callback).toStrictEqual(
        expect.any(Function),
      );
    });
  });

  describe("when registering an event handler to multiple existing contexts", () => {
    let domain: ViewDomain<any>;

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: ["context", "secondary"],
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("eventName", ctx.event.name);
          },
        }),
      );
    });

    it("should throw", async () => {
      await expect(
        domain.registerEventHandler(
          new ViewEventHandler({
            aggregate: {
              name: "name",
              context: ["context", "secondary"],
            },
            view: {
              name: "name",
              context: "context",
            },
            eventName: "eventName",
            getViewId: (event) => event.aggregate.id,
            handler: async (ctx) => {
              ctx.setState("eventName", ctx.event.name);
            },
          }),
        ),
      ).rejects.toThrow(Error);
    });

    it("should become part of the domain only once for each context", () => {
      // @ts-ignore
      expect(domain.eventHandlers).toMatchSnapshot();
    });
  });

  describe("when handling an event", () => {
    let domain: ViewDomain<any>;
    const emitSpy = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      domain.on("error", (error) => {
        emitSpy({ error });
      });

      domain.on("event", (error, event, view) => {
        emitSpy({ error, event, view });
      });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
          },
        }),
      );
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();
    });

    it("should save the view to the view store", () => {
      // @ts-ignore
      expect(domain.viewStore.store.views).toMatchSnapshot();
    });

    it("should emit the handled event and view", () => {
      expect(emitSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe("when handling an event twice", () => {
    let domain: ViewDomain<any>;
    const handlerSpy = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
            handlerSpy();
          },
        }),
      );
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();

      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();
    });

    it("should save only one view to the view store", () => {
      // @ts-ignore
      expect(domain.viewStore.store.views).toMatchSnapshot();
    });

    it("should call the event handler only once", () => {
      expect(handlerSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe("when handling an event without a matching event handler", () => {
    let domain: ViewDomain<any>;

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });
    });

    it("should reject", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).rejects.toThrow(HandlerNotRegisteredError);
    });
  });

  describe("when failing to handle an event", () => {
    let domain: ViewDomain<any>;
    const emitSpy = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      domain.on("error", (error) => {
        emitSpy({ error });
      });

      domain.on("event", (error, event, view) => {
        emitSpy({ error, event, view });
      });

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            throw new Error("handler error");
          },
        }),
      );
    });

    it("should reject", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).rejects.toThrow(Error);
    });

    it("should not save the view", () => {
      // @ts-ignore
      expect(domain.viewStore.store.views).toStrictEqual({});
    });

    it("should emit the error and not the event", () => {
      expect(emitSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe("when failing to handle an event with a non-permanent domain error", () => {
    let domain: ViewDomain<any>;

    class TestError extends DomainError {
      public constructor() {
        super("handler error", { permanent: false });
      }
    }

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            throw new TestError();
          },
        }),
      );
    });

    it("should reject", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).rejects.toThrow(TestError);
    });

    it("should not save the view", () => {
      // @ts-ignore
      expect(domain.viewStore.store.views).toStrictEqual({});
    });
  });

  describe("when failing to handle an event with a permanent domain error", () => {
    let domain: ViewDomain<any>;

    class TestError extends DomainError {
      public constructor() {
        super("handler error", { permanent: true });
      }
    }

    beforeAll(async () => {
      domain = new ViewDomain<any>({ messageBus: {}, viewStore: {}, logger });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            throw new TestError();
          },
        }),
      );
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();
    });

    it("should not save the view", () => {
      // @ts-ignore
      expect(domain.viewStore.store.views).toStrictEqual({});
    });

    it("should publish non-mandatory error event to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toMatchSnapshot();
    });
  });

  describe("when handling an event on a destroyed domain", () => {
    let domain: ViewDomain<any>;
    const spySave = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({
        messageBus: {},
        viewStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async (viewIdentifier) => {
              return new View({ ...viewIdentifier, destroyed: true }, logger);
            },
            save: async (view) => {
              spySave(view);
              return new View({ ...view, revision: view.revision + 1 }, logger);
            },
            query: async () => {
              return [new View({ id: "id", name: "name", context: "context" }, logger)];
            },
          },
        },
        logger,
      });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
          },
        }),
      );
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();
    });

    it("should not save the view", () => {
      expect(spySave).not.toHaveBeenCalled();
    });

    it("should publish ViewAlreadyDestroyedError to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toMatchSnapshot();
    });
  });

  describe("when handling an event with a created [true] condition", () => {
    let domain: ViewDomain<any>;
    const spySave = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({
        messageBus: {},
        viewStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async () => {
              return null;
            },
            save: async (view) => {
              spySave(view);
              return new View({ ...view, revision: view.revision + 1 }, logger);
            },
            query: async () => {
              return [new View({ id: "id", name: "name", context: "context" }, logger)];
            },
          },
        },
        logger,
      });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          conditions: {
            created: true,
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
          },
        }),
      );
    });

    it("should reject with a transient ViewNotCreatedError", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).rejects.toThrow(ViewNotCreatedError);
    });
  });

  describe("when handling an event with a created [true] and permanent [true] condition", () => {
    let domain: ViewDomain<any>;
    const spySave = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({
        messageBus: {},
        viewStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async () => {
              return null;
            },
            save: async (view) => {
              spySave(view);
              return new View({ ...view, revision: view.revision + 1 }, logger);
            },
            query: async () => {
              return [new View({ id: "id", name: "name", context: "context" }, logger)];
            },
          },
        },
        logger,
      });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          conditions: {
            created: true,
            permanent: true,
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
          },
        }),
      );
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();
    });

    it("should not save the view", () => {
      expect(spySave).not.toHaveBeenCalled();
    });

    it("should publish ViewNotCreatedError to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toMatchSnapshot();
    });
  });

  describe("when handling an event with a created [false] condition", () => {
    let domain: ViewDomain<any>;
    const spySave = jest.fn();

    beforeAll(async () => {
      domain = new ViewDomain<any>({
        messageBus: {},
        viewStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async (viewIdentifier) => {
              return new View({ ...viewIdentifier, revision: 1 }, logger);
            },
            save: async (view) => {
              spySave(view);
              return new View({ ...view, revision: view.revision + 1 }, logger);
            },
            query: async () => {
              return [new View({ id: "id", name: "name", context: "context" }, logger)];
            },
          },
        },
        logger,
      });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          conditions: {
            created: false,
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
          },
        }),
      );
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).resolves.toBeUndefined();
    });

    it("should not save the view", () => {
      expect(spySave).not.toHaveBeenCalled();
    });

    it("should publish ViewAlreadyCreatedError to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toMatchSnapshot();
    });
  });

  describe("when handling an event with a created [false] and permanent [false] condition", () => {
    let domain: ViewDomain<any>;

    beforeAll(async () => {
      domain = new ViewDomain<any>({
        messageBus: {},
        viewStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async (viewIdentifier) => {
              return new View({ ...viewIdentifier, revision: 1 }, logger);
            },
            save: async (view) => {
              return new View({ ...view, revision: view.revision + 1 }, logger);
            },
            query: async () => {
              return [new View({ id: "id", name: "name", context: "context" }, logger)];
            },
          },
        },
        logger,
      });

      domain.on("error", () => {});
      domain.on("event", () => {});

      await domain.registerEventHandler(
        new ViewEventHandler({
          aggregate: {
            name: "name",
            context: "context",
          },
          conditions: {
            created: false,
            permanent: false,
          },
          view: {
            name: "name",
            context: "context",
          },
          eventName: "eventName",
          getViewId: (event) => event.aggregate.id,
          handler: async (ctx) => {
            ctx.setState("data", ctx.event.data);
          },
        }),
      );
    });

    it("should reject with a transient ViewAlreadyCreatedError", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(
          new DomainEvent({
            aggregate: {
              id: "id",
              name: "name",
              context: "context",
            },
            data: { data: "data" },
            name: "eventName",
          }),
          {
            name: "name",
            context: "context",
          },
        ),
      ).rejects.toThrow(ViewAlreadyCreatedError);
    });
  });
});
