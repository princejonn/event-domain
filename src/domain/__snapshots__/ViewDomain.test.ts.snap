// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewDomain when failing to handle an event should emit the error and not the event 1`] = `
Array [
  Array [
    Object {
      "error": [Error: handler error],
    },
  ],
]
`;

exports[`ViewDomain when failing to handle an event with a permanent domain error should publish non-mandatory error event to message bus 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "context",
      "id": "id",
      "name": "name",
    },
    "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {
      "error": [TestError: handler error],
      "message": DomainEvent {
        "aggregate": Object {
          "context": "context",
          "id": "id",
          "name": "name",
        },
        "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "data": Object {
          "data": "data",
        },
        "delay": 0,
        "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "mandatory": false,
        "name": "eventName",
        "timestamp": 2022-01-01T08:00:00.000Z,
        "type": "event",
      },
    },
    "delay": 0,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1.TestError",
    "mandatory": false,
    "name": "TestError",
    "timestamp": 2022-01-01T08:00:00.000Z,
    "type": "event",
  },
]
`;

exports[`ViewDomain when handling an event on a destroyed domain should publish ViewAlreadyDestroyedError to message bus 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "context",
      "id": "id",
      "name": "name",
    },
    "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {
      "error": [ViewDestroyedError: View is destroyed],
      "message": DomainEvent {
        "aggregate": Object {
          "context": "context",
          "id": "id",
          "name": "name",
        },
        "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "data": Object {
          "data": "data",
        },
        "delay": 0,
        "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "mandatory": false,
        "name": "eventName",
        "timestamp": 2022-01-01T08:00:00.000Z,
        "type": "event",
      },
    },
    "delay": 0,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1.ViewDestroyedError",
    "mandatory": false,
    "name": "ViewDestroyedError",
    "timestamp": 2022-01-01T08:00:00.000Z,
    "type": "event",
  },
]
`;

exports[`ViewDomain when handling an event should emit the handled event and view 1`] = `
Array [
  Array [
    Object {
      "error": null,
      "event": DomainEvent {
        "aggregate": Object {
          "context": "context",
          "id": "id",
          "name": "name",
        },
        "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "data": Object {
          "data": "data",
        },
        "delay": 0,
        "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "mandatory": false,
        "name": "eventName",
        "timestamp": 2022-01-01T08:00:00.000Z,
        "type": "event",
      },
      "view": Object {
        "causationList": Array [
          "e64b576a-022e-4dd4-8c97-32408012f8d1",
        ],
        "context": "context",
        "destroyed": false,
        "id": "id",
        "metaState": Object {
          "data": Object {
            "r": false,
            "t": 2022-01-01T08:00:00.000Z,
            "v": Object {
              "data": "data",
            },
          },
        },
        "name": "name",
        "revision": 0,
        "state": Object {
          "data": Object {
            "data": "data",
          },
        },
      },
    },
  ],
]
`;

exports[`ViewDomain when handling an event should save the view to the view store 1`] = `
Object {
  "context.name.id": Object {
    "causationList": Array [
      "e64b576a-022e-4dd4-8c97-32408012f8d1",
    ],
    "context": "context",
    "destroyed": false,
    "id": "id",
    "metaState": Object {
      "data": Object {
        "r": false,
        "t": 2022-01-01T08:00:00.000Z,
        "v": Object {
          "data": "data",
        },
      },
    },
    "name": "name",
    "revision": 1,
    "state": Object {
      "data": Object {
        "data": "data",
      },
    },
  },
}
`;

exports[`ViewDomain when handling an event twice should save only one view to the view store 1`] = `
Object {
  "context.name.id": Object {
    "causationList": Array [
      "e64b576a-022e-4dd4-8c97-32408012f8d1",
    ],
    "context": "context",
    "destroyed": false,
    "id": "id",
    "metaState": Object {
      "data": Object {
        "r": false,
        "t": 2022-01-01T08:00:00.000Z,
        "v": Object {
          "data": "data",
        },
      },
    },
    "name": "name",
    "revision": 1,
    "state": Object {
      "data": Object {
        "data": "data",
      },
    },
  },
}
`;

exports[`ViewDomain when handling an event with a created [false] condition should publish ViewAlreadyCreatedError to message bus 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "context",
      "id": "id",
      "name": "name",
    },
    "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {
      "error": [ViewAlreadyCreatedError: View has already been created],
      "message": DomainEvent {
        "aggregate": Object {
          "context": "context",
          "id": "id",
          "name": "name",
        },
        "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "data": Object {
          "data": "data",
        },
        "delay": 0,
        "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "mandatory": false,
        "name": "eventName",
        "timestamp": 2022-01-01T08:00:00.000Z,
        "type": "event",
      },
    },
    "delay": 0,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1.ViewAlreadyCreatedError",
    "mandatory": false,
    "name": "ViewAlreadyCreatedError",
    "timestamp": 2022-01-01T08:00:00.000Z,
    "type": "event",
  },
]
`;

exports[`ViewDomain when handling an event with a created [true] and permanent [true] condition should publish ViewNotCreatedError to message bus 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "context",
      "id": "id",
      "name": "name",
    },
    "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {
      "error": [ViewNotCreatedError: View has not been created],
      "message": DomainEvent {
        "aggregate": Object {
          "context": "context",
          "id": "id",
          "name": "name",
        },
        "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "data": Object {
          "data": "data",
        },
        "delay": 0,
        "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
        "mandatory": false,
        "name": "eventName",
        "timestamp": 2022-01-01T08:00:00.000Z,
        "type": "event",
      },
    },
    "delay": 0,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1.ViewNotCreatedError",
    "mandatory": false,
    "name": "ViewNotCreatedError",
    "timestamp": 2022-01-01T08:00:00.000Z,
    "type": "event",
  },
]
`;

exports[`ViewDomain when registering an event handler should add event subscription 1`] = `
Object {
  "aggregate": Object {
    "context": "context",
    "name": "name",
  },
  "handler": [Function],
  "name": "context.name.eventName",
}
`;

exports[`ViewDomain when registering an event handler should add subscriptions to topic 1`] = `
Object {
  "aggregate": Object {
    "context": "context",
    "name": "name",
  },
  "handler": [Function],
  "name": "eventName",
}
`;

exports[`ViewDomain when registering an event handler should become part of the domain 1`] = `
Array [
  ViewEventHandler {
    "aggregate": Object {
      "context": "context",
      "name": "name",
    },
    "conditions": Object {},
    "eventName": "eventName",
    "getViewId": [Function],
    "handler": [Function],
    "loadOptions": Object {},
    "saveOptions": Object {},
    "view": Object {
      "context": "context",
      "name": "name",
    },
  },
]
`;

exports[`ViewDomain when registering an event handler to multiple contexts should add event subscription 1`] = `
Object {
  "aggregate": Object {
    "context": "context",
    "name": "name",
  },
  "handler": [Function],
  "name": "context.name.eventName",
}
`;

exports[`ViewDomain when registering an event handler to multiple contexts should add subscriptions to topic 1`] = `
Object {
  "aggregate": Object {
    "context": "context",
    "name": "name",
  },
  "handler": [Function],
  "name": "eventName",
}
`;

exports[`ViewDomain when registering an event handler to multiple contexts should become part of the domain for each context 1`] = `
Array [
  ViewEventHandler {
    "aggregate": Object {
      "context": "context",
      "name": "name",
    },
    "conditions": Object {},
    "eventName": "eventName",
    "getViewId": [Function],
    "handler": [Function],
    "loadOptions": Object {},
    "saveOptions": Object {},
    "view": Object {
      "context": "context",
      "name": "name",
    },
  },
  ViewEventHandler {
    "aggregate": Object {
      "context": "secondary",
      "name": "name",
    },
    "conditions": Object {},
    "eventName": "eventName",
    "getViewId": [Function],
    "handler": [Function],
    "loadOptions": Object {},
    "saveOptions": Object {},
    "view": Object {
      "context": "context",
      "name": "name",
    },
  },
]
`;

exports[`ViewDomain when registering an event handler to multiple existing contexts should become part of the domain only once for each context 1`] = `
Array [
  ViewEventHandler {
    "aggregate": Object {
      "context": "context",
      "name": "name",
    },
    "conditions": Object {},
    "eventName": "eventName",
    "getViewId": [Function],
    "handler": [Function],
    "loadOptions": Object {},
    "saveOptions": Object {},
    "view": Object {
      "context": "context",
      "name": "name",
    },
  },
  ViewEventHandler {
    "aggregate": Object {
      "context": "secondary",
      "name": "name",
    },
    "conditions": Object {},
    "eventName": "eventName",
    "getViewId": [Function],
    "handler": [Function],
    "loadOptions": Object {},
    "saveOptions": Object {},
    "view": Object {
      "context": "context",
      "name": "name",
    },
  },
]
`;

exports[`ViewDomain when registering an existing event handler should not register a second handler on domain 1`] = `
Array [
  ViewEventHandler {
    "aggregate": Object {
      "context": "context",
      "name": "name",
    },
    "conditions": Object {},
    "eventName": "eventName",
    "getViewId": [Function],
    "handler": [Function],
    "loadOptions": Object {},
    "saveOptions": Object {},
    "view": Object {
      "context": "context",
      "name": "name",
    },
  },
]
`;
